src/mlpack/core/tree/binary_space_tree/ub_tree_split_impl.hpp

```diff
@@ -0,0 +1,244 @@
+  }
+
+  // Set the minimum and the maximum addresses.
+  for (size_t k = 0; k < bound.Dim(); k++)
+  {
+    bound.LoAddress()[k] = addresses[begin].first[k];
+    bound.HiAddress()[k] = addresses[begin + count - 1].first[k];
+  }
+  bound.UpdateAddressBounds(data.cols(begin, begin + count - 1));
+
+  return true;
+}
+
+template<typename BoundType, typename MatType>
+void UBTreeSplit<BoundType, MatType>::InitializeAddresses(const MatType& data)
+{
+  addresses.resize(data.n_cols);
+
+  // Calculate all addresses.
+  for (size_t i = 0; i < data.n_cols; i++)
+  {
+    addresses[i].first.zeros(data.n_rows);
+    bound::addr::PointToAddress(addresses[i].first, data.col(i));
+    addresses[i].second = i;
+  }
+}
+
+template<typename BoundType, typename MatType>
+size_t UBTreeSplit<BoundType, MatType>::PerformSplit(
+    MatType& data,
+    const size_t begin,
+    const size_t count,
+    const SplitInfo& splitInfo)
+{
+  // For the first time we have to rearrange the dataset.
+  if (splitInfo.addresses)
+  {
+    std::vector<size_t> newFromOld(data.n_cols);
+    std::vector<size_t> oldFromNew(data.n_cols);
+
+    for (size_t i = 0; i < splitInfo.addresses->size(); i++)
+    {
+      newFromOld[i] = i;
+      oldFromNew[i] = i;
+    }
+
+    for (size_t i = 0; i < splitInfo.addresses->size(); i++)
+    {
+      size_t index = (*splitInfo.addresses)[i].second;
+      size_t oldI = oldFromNew[i];
+      size_t newIndex = newFromOld[index];
+
+      data.swap_cols(i, newFromOld[index]);
+
+      size_t tmp = newFromOld[index];
+      newFromOld[index] = i;
+      newFromOld[oldI] = tmp;
+
+      tmp = oldFromNew[i];
+      oldFromNew[i] = oldFromNew[newIndex];
+      oldFromNew[newIndex] = tmp;
+    }
+  }
+
+  // Since the dataset is sorted we can easily obtain the split column.
+  return begin + count / 2;
+}
+
+template<typename BoundType, typename MatType>
+size_t UBTreeSplit<BoundType, MatType>::PerformSplit(
+    MatType& data,
+    const size_t begin,
+    const size_t count,
+    const SplitInfo& splitInfo,
+    std::vector<size_t>& oldFromNew)
+{
+  // For the first time we have to rearrange the dataset.
+  if (splitInfo.addresses)
+  {
+    std::vector<size_t> newFromOld(data.n_cols);
+    std::vector<size_t> newToOld(data.n_cols);
+
+    for (size_t i = 0; i < splitInfo.addresses->size(); i++)
+    {
+      newFromOld[i] = i;
+      newToOld[i] = i;
+    }
+
+    for (size_t i = 0; i < splitInfo.addresses->size(); i++)
+    {
+      size_t index = (*splitInfo.addresses)[i].second;
+      size_t oldI = newToOld[i];
+      size_t newIndex = newFromOld[index];
+
+      data.swap_cols(i, newFromOld[index]);
+
+      size_t tmp = newFromOld[index];
+      newFromOld[index] = i;
+      newFromOld[oldI] = tmp;
+
+      tmp = newToOld[i];
+      newToOld[i] = newToOld[newIndex];
+      newToOld[newIndex] = tmp;
+
+      tmp = oldFromNew[i];
+      oldFromNew[i] = oldFromNew[newIndex];
+      oldFromNew[newIndex] = tmp;
```

Comment:
Yeah; First time I had used `oldFromNew` instead of `newToOld`. But then I had thought that the dataset can be rearranged before the tree construction (that never happens right now).